sebs/knative/knative.py

```diff
@@ -0,0 +1,552 @@
+from typing import Dict, Tuple, Type, List, Optional
+import docker
+from .function import KnativeFunction, KnativeFunctionConfig
+from typing import cast
+from .config import KnativeConfig
+
+
+class Knative(System):
+    _config: KnativeConfig
+
+    def __init__(
+        self,
+        system_config: SeBSConfig,
+        config: KnativeConfig,
+        cache_client: Cache,
+        docker_client: docker.client,
+        logger_handlers: LoggingHandlers,
+    ):
+        super().__init__(system_config, cache_client, docker_client)
+        self._config = config
+        self._logging_handlers = logger_handlers
+
+        if self.config.resources.docker_username:
+            if self.config.resources.docker_registry:
+                docker_client.login(
+                    username=self.config.resources.docker_username,
+                    password=self.config.resources.docker_password,
+                    registry=self.config.resources.docker_registry,
+                )
+            else:
+                docker_client.login(
+                    username=self.config.resources.docker_username,
+                    password=self.config.resources.docker_password,
+                )
+
+    def initialize(
+        self, config: Dict[str, str] = {}, resource_prefix: Optional[str] = None
+    ):
+        self.initialize_resources(select_prefix=resource_prefix)
+
+    @property
+    def config(self) -> KnativeConfig:
+        return self._config
+
+    def get_storage(self, replace_existing: bool = False) -> PersistentStorage:
+        if not hasattr(self, "storage"):
+
+            if not self.config.resources.storage_config:
+                raise RuntimeError(
+                    "Knative is missing the configuration of pre-allocated storage!"
+                )
+            self.storage = KnativeMinio.deserialize(
+                self.config.resources.storage_config,
+                self.cache_client,
+                self.config.resources,
+            )
+            self.storage.logging_handlers = self.logging_handlers
+        else:
+            self.storage.replace_existing = replace_existing
+        return self.storage
+
+    def shutdown(self) -> None:
+        if hasattr(self, "storage") and self.config.shutdownStorage:
+            self.storage.stop()
+        if self.config.removeCluster:
+            from tools.knative_preparation import delete_cluster
+
+            delete_cluster()
+        super().shutdown()
+
+    @staticmethod
+    def name() -> str:
+        return "knative"
+
+    @staticmethod
+    def typename():
+        return "Knative"
+
+    @staticmethod
+    def function_type() -> Type[Function]:
+        return KnativeFunction
+
+    def get_knative_func_cmd(self) -> List[str]:
+        cmd = [self.config.knative_exec]
+        return cmd
+
+    def find_image(self, repository_name: str, image_tag: str) -> bool:
+        """
+        Check if a Docker image exists in the repository.
```

Comment:
We should replace it with a version that does not pull image if we don't have to use it locally. Otherwise, we waste gigabytes of network traffic for nothing. Please check the code from OpenWhisk; there's a feature for that.